lib/cylc/remote.py

```diff
@@ -18,15 +18,79 @@
 """Run command on a remote, (i.e. a remote [user@]host)."""
 
 import os
-from posix import WIFSIGNALED
-from pipes import quote
-import shlex
-from subprocess import Popen
 import sys
-from textwrap import TextWrapper
+import shlex
+from pipes import quote
+from posix import WIFSIGNALED
+
+# CODACY ISSUE:
+#   Consider possible security implications associated with Popen module.
+# REASON IGNORED:
+#   Subprocess is needed, but we use it with security in mind.
+from subprocess import Popen, PIPE
 
-from cylc.cfgspec.glbl_cfg import glbl_cfg
 import cylc.flags
+from cylc.cfgspec.glbl_cfg import glbl_cfg
+from cylc.version import CYLC_VERSION
+
+
+def remote_cylc_cmd(cmd, user=None, host=None, capture=False,
+                    ssh_login_shell=None):
+    """Run a given cylc command on another account and/or host.
+
+    """
+    if host is None:
+        host = "localhost"
+    if user is None:
+        user_at_host = host
+    else:
+        user_at_host = "%s@%s" % (user, host)
+
+    # Pass cylc version through.
+    command = ["env", "CYLC_VERSION=%s" % CYLC_VERSION]
+
+    ssh = str(glbl_cfg().get_host_item("ssh command", host, user))
+    command += shlex.split(ssh) + ["-n", user_at_host]
+
+    # Use bash loging shell?
+    if ssh_login_shell is None:
+        ssh_login_shell = glbl_cfg().get_host_item(
+            "use login shell", host, user)
+    if ssh_login_shell:
+        # A login shell will always source /etc/profile and the user's bash
+        # profile file. To avoid having to quote the entire remote command
+        # it is passed as arguments to bash.
+        command += ["bash", "--login", "-c", "'exec $0 \"$@\"'"]
```

Comment:
Neater to do `quote(r'exec "$0" "$@"')`?

Reply:
I think I just took this from the existing code in `remote.py`, but OK...

Reply:
Sorry, yes you did. I've only changed it in #2582.